src/main/java/com/google/devtools/build/bfg/JavaSourceFileParser.java

```diff
@@ -124,11 +201,45 @@
         .forEach(
             (path, classes) -> {
               if (any(oneRulePerPackageRoots, p -> path.startsWith(p))) {
-                putOnCycle(classes, graph);
+                putOnCycle(classes, classToClass);
               }
             });
+  }
+
+  private String decideRuleKind(ReferencedClassesParser parser, Set<String> dependencies) {
```

Comment:
nit: can you separate this out to another class?

This heuristic functionality is independent enough that it should be in its own class, with tests of course. Doing this makes adding support for Scala and kotlin much cleaner: you would just add another class for making these decisions. 

Also this is neat!

Reply:
> To be honest I think there are too many class in BFG as is :)
...and not enough packages! :)

> Scala/Kotlin will have to write their own parser anyway, with their own heuristics and own return values, so I don't see the reuse value.

It slipped my mind they would have their own parsers. With that in mind, dismiss my previous suggestion.